internal/controllers/addoninstance_controller.go

```diff
@@ -0,0 +1,62 @@
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/openshift/reference-addon/internal/addonsdk"
+)
+
+type AddonInstanceReconciler struct {
+	client.Client
+	StatusReporter  *addonsdk.StatusReporter
+	Log             logr.Logger
+	TargetNamespace string
+}
+
+func (r *AddonInstanceReconciler) Reconcile(
+	ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	addonInstance := &addonsv1alpha1.AddonInstance{}
+	if err := r.Get(ctx, req.NamespacedName, addonInstance); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	r.Log.Info("reporting addon-instance spec change to status-reporter")
+	if err := r.StatusReporter.ReportAddonInstanceSpecChange(ctx, addonInstance); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *AddonInstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	addonInstanceConfigurationChangePredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Object.GetName() == "addon-instance" && e.Object.GetNamespace() == r.TargetNamespace
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// AddonInstance CR to be watched should be only <target-namespace>/addon-instance
+			// ignore updates to .status in which case metadata.generation does not change
+			return e.ObjectNew.GetName() == "addon-instance" && e.ObjectNew.GetNamespace() == r.TargetNamespace && e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
```

Comment:
To filter for a specific name, we should rather do the same as in the `ReferenceAddonReconciler` and just check `req.NamespacedName.Name != "reference-addon"` at the start of the `Reconcile` function.
Namespace filtering should not be done here, but instead by limiting the whole manager to a specific namespace:
https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.1/pkg/manager#Options (Namespace)

Or by setting up a custom cache reader to make sure `AddonInstance` objects are only cached for a specific namespace. Otherwise we have to give every Addon cluster wide list/watch permissions on `AddonInstance` objects, which will make it very likely that an Addon is accidentally reconciling on other AddonInstance objects when multiple Addons are installed.

Reply:
As per our discussion, I configured the manager to only cache the objects reisiding in targetNamespace considering the non-triviality associated with configuring it only cache AddonInstance CR objects in the targetNamespace and other objects cluster-wide instead.